proposals/0000-exception-backtraces.rst

```diff
@@ -0,0 +1,678 @@
+compile-time error complaining that ``?exceptionContext`` is not in scope.
+In our experience, this sort of code is rare and generally quite
+straightforward to adapt; a survey of Hackage suggests that nearly all uses of
+``SomeException`` are in pattern contexts.
+
+We expect that users relying on exceptions (in particular asychronous
+exceptions) to adjust control flow in non-exceptional situations (e.g.
+cancellation in the ``async`` package) will want to
+define ``backtraceDesired = False`` in their ``Exception`` instances.
+
+
+Alternatives
+------------
+
+Exception hierarchy design (alternative one)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+An earlier version of this proposal changed the root of the exception hierarchy
+to a new type which included a backtrace: ::
+
+    data SomeExceptionWithBacktrace
+      = SomeExceptionWithBacktrace
+          :: SomeException       -- ^ the exception
+          -> [Backtrace]         -- ^ backtraces
+          -> SomeExceptionWithBacktrace
+
+Unsurprisingly, this change had a non-negligible
+impact on existing user code. Moreover, the
+change introduced confusion as users of the old
+``SomeException`` type would silently not benefit from the
+introduction of backtraces. Moreover, this proposal was
+considerably less generic, focusing on static backtraces
+instead of arbitrary user-defined annotations.
+
+Exception hierarchy design (alternative two)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Yet an earlier version suggested keeping ``SomeException`` as the root exception
+type, changing the constructor to add a ``Maybe Backtrace`` field and a pattern
+synonym for backwards compatibility: ::
+
+    data SomeException where
+      SomeExceptionWithLocation
+        :: forall e. Exception e
+        => Maybe Backtrace   -- ^ backtrace, if available
+        -> e                 -- ^ the exception
+        -> SomeException
+
+    pattern SomeException e <- SomeExceptionWithLocation _ e
+      where
+        SomeException e = mkSomeExceptionWithLocation e
+
+The problem with this is that the pattern match completeness checker does not
+play well with pattern synonyms. Additionally, it may introduce a ``MonadFail``
+constraint where one previously did not exist. For example, the following would no
+longer typecheck due to the lack of a ``MonadFail m`` constraint: ::
+
+    f :: Monad m => SomeException -> m ()
+    f someException = do
+      SomeException e <- pure someException   -- Pattern synonym is assumed fallible
+      ...
+
+Backtrace mechanism selection
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+In addition, there are several alternatives to the proposed backtrace mechanism
+selection facility. For instance:
+
+* a simpler, non-GADT-based approach might be used
+* GHC could gain support for setting the backtrace mechanism at compile-time
+  via a compiler flag (this would essentially come down to GHC emitting a call
+  to ``enabledBacktraceMechanisms`` in its start-up code).
```

Comment:
Why not have GHC default to enabling all backtrace mechanisms that might contain information? I think with the exception of `HasCallStack` they're all enabled by compiler flags, so it should be easy for GHC to know which mechanisms to enable.

I can see a perf concern, but between

1. disabling `backtraceDesired` for control-flow exceptions, and
2. the fact that the expensive mechanisms seem unlikely to be enabled outside of specific debugging builds

the concern seems surmountable to me.

Reply:
> so it should be easy for GHC to know which mechanisms to enable.

Unfortunately this is true only for the cost-centre and `HasCallStack` mechanisms. Both IPE and DWARF are trickier in that they are not "global": some modules may be built with IPE/DWARF information while others may not.

To make matters worse, IPE mechanism performs its metadata lookups lazily: it freezes the stack when the exception is thrown (essentially performing a potentially expensive copy) and only tries to lookup the associated IPE entries when someone tries to examine the backtrace. This means that users enabling the IPE mechanism will end up incurring the cost of a stack copy regardless of whether they have IPE information available. Technically we could use heuristics (e.g. checking whether the IPE table has no entries) to disable IPE backtrace collection but I'm reluctant to do this since they would be hard to predict and easily defeated (e.g. if your compiler was built with IPE information but your application was not).